deps/picotls/lib/picotls.c

```diff
@@ -5628,9 +5652,159 @@ char *ptls_hexdump(char *buf, const void *_src, size_t len)
     size_t i;
 
     for (i = 0; i != len; ++i) {
-        *dst++ = "0123456789abcdef"[src[i] >> 4];
-        *dst++ = "0123456789abcdef"[src[i] & 0xf];
+        ptls_byte_to_hex(dst, src[i]);
+        dst += 2;
     }
     *dst++ = '\0';
     return buf;
 }
+
+/**
+ * Builds a JSON-safe string without double quotes. Supplied buffer MUST be 6x + 1 bytes larger than the input.
+ */
+static size_t escape_json_unsafe_string(char *buf, const void *unsafe_str, size_t len)
+{
+    char *dst = buf;
+    const uint8_t *src = unsafe_str, *end = src + len;
+
+    for (; src != end; ++src) {
+        switch (*src) {
```

Comment:
FWIW the JSON serialization algorithm comes from picojson: 

https://github.com/kazuho/picojson/blob/111c9be5188f7350c2eac9ddaedd8cca3d7bf394/picojson.h#L531-L556